lmcache/blend/retriever.py

```diff
@@ -214,60 +214,73 @@ def __init__(
-                                                             spt_len])
-            start = i + spt_len
-
-        if start < len(input_tokens_single_query):
-            splitted_tokens.append(input_tokens_single_query[start:])
-        return splitted_tokens
+        start = 0
+        while True:
+            next_len = i + spt_len
+            if next_len > len(full_prompt):
+                break
+            if full_prompt[i:next_len] == self.spt:
+                splitted_tokens.append(full_prompt[start:i])
+                start = next_len
+                i = next_len
+            else:
+                i += 1
+
+        if start < len(full_prompt):
+            splitted_tokens.append(full_prompt[start:])
+
+        new_prompt = []
+        new_indices = [0]
+        for split in splitted_tokens:
+            new_prompt.extend(split)
+            new_indices.append(new_indices[-1] + len(split))
+        new_indices[-1] = len(new_prompt)
+        return new_prompt, new_indices
 
     def new_request(
         self,
-        input_tokens: torch.Tensor,
-        query_start_loc: torch.Tensor,
+        full_prompts: List[torch.Tensor],
+        indices: List[List[int]],
     ) -> BlendRetrieverTask:
         """Create a new BlendRetrieverTask to retrieve the KV caches.
         It may launch async tasks in the background during the retrieval.
 
-        :param torch.Tensor input_tokens: The input tokens, could include
-            multiple requests in a batch
-        :param torch.Tensor query_start_loc: The start location of the query if
-            input_tokens has multiple requests in a batch. The length should be 
-            the number of requests in the batch + 1.
+        :param List[torch.Tensor] full_prompts: The full prompts for each
+        request in this batch.
+        :param List[List[int]] indices: The indices of where the 
+        segmengted requests start in the full prompts.
 
         :return: The retriever task to retrieve the KV caches
         :rtype: BlendRetrieverTask
         """
+        assert len(full_prompts) == len(indices)
         with ThreadPoolExecutor(max_workers=1) as executor:
             splitted_tokens = []
-            start_loc = query_start_loc[0]
-            for loc in query_start_loc[1:]:
-                logger.debug(f"Request start loc = {start_loc}")
-                splitted_tokens.extend(
-                    self._split_input_tokens(input_tokens[start_loc:loc]))
-                start_loc = loc
-
+            for prompt_idx, prompt in enumerate(full_prompts):
+                prompt_indices = indices[prompt_idx]
+                assert len(prompt_indices) >= 2
+                segment_num = len(prompt_indices) - 1
+                for i in range(segment_num):
+                    start_loc = prompt_indices[i]
+                    end_loc = prompt_indices[i + 1]
+                    if end_loc > start_loc:
+                        splitted_tokens.append(prompt[start_loc:end_loc])
```

Comment:
Same as above. Can we use as much tensor operations as possible?

Reply:
> Same as above. Can we use as much tensor operations as possible?

The indices part is possible to improve, but I think we will not cut it into too many chunks, probably in 10 in one prompt.